doc/manual/contract.md

```diff
@@ -0,0 +1,421 @@
+	end
+
+	class MySub
+		super MyClass
+
+		redef fun foo(x: Int) is expect(x == -1)
+	end
+
+	class MySubSub
+		super MySub
+
+		redef fun foo(x: Int) is expect(x == -2)
+	end
+~~~
+
+| Class    | Condition of `foo(x)`          |
+|----------|--------------------------------|
+| MyClass  | `x == 0`                       |
+| MySub    | `x == 0 or x == -1`            |
+| MySubSub | `x == 0 or x == -1 or x == -2` |
+
+When refinement a precondition (`expect`) the new condition is equivalent to an `or` between the set of inherited preconditions.
+
+* The `ensure` can be reinforced, which guarantees that all of the expected results introduced in the previous definitions must be respected. During the specialization of a method it will be possible to redefine its postcondition in order to restrict the possible outputs.
+
+Example:
+~~~
+	class MyClass
+		fun foo: Int is abstract, ensure(result > 0)
+	end
+
+	class MySub
+		super MyClass
+
+		redef fun foo: Int is abstract, ensure(result > 10)
+	end
+
+	class MySubSub
+		super MyClass
+
+		redef fun foo: Int is abstract, ensure(result > 100)
+	end
+~~~
+
+| Class    | Condition of `foo`                            |
+|----------|-----------------------------------------------|
+| MyClass  | `result > 0`                                  |
+| MySub    | `result > 0 and result > 10`                  |
+| MySubSub | `result > 0 and result > 10 and result > 100` |
+
+When refinement a postcondition (`ensure`) the new condition is equivalent to an `and` between the set of inherited postconditions.
+
+* The `invariant` can be reinforced, which guarantees that constraints of validated state previously defined in the superclasses/interfaces must be respected. The specialization of a class makes it possible to restrict all of the valid states. The objective is that an object must always be considered as a coherent value of all of its superclasses/interfaces.
+
+Example:
+~~~
+	class MyClass
+		invariant(i > 0)
+		var i: Int
+	end
+
+	class MySub
+		super MyClass
+		invariant(i > 10)
+	end
+
+	class MySubSub
+		super MySub
+		invariant(i > 100)
+	end
+~~~
+
+| Class    | Condition of `invariant`       |
+|----------|--------------------------------|
+| MyClass  | `i > 0`                        |
+| MySub    | `i > 0 and i > 10`             |
+| MySubSub | `i > 0 and i > 10 and i > 100` |
+
+When we specialize or refine a class, its invariant is equal to an `and` between the set of inherited invariants and the new condition.
+
+Nit however offers the possibility of removing the inheritance of contracts with the `no_contract`  annotation (use this annotation only when it's really necessary).
+
+Example:
+~~~
+	class MyClass
+		fun foo(x: Int) is abstract, ensure(x == 0)
+	end
+
+	class MySub
+		super MyClass
+
+		redef fun foo(x: Int) is no_contract
+	end
+~~~
+
+| Class   | Condition of `foo(x)` |
+|---------|-----------------------|
+| MyClass | `i == 0`              |
+| MySub   | `Null`                |
```

Comment:
`Null` is not a real condition. Maybe use *none* (without code formatting) or `true` ?